modules/040-node-manager/hooks/migration/migrate_system_reserve.go

```diff
@@ -0,0 +1,123 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package hooks
+
+import (
+	"github.com/flant/addon-operator/pkg/module_manager/go_hook"
+	"github.com/flant/addon-operator/sdk"
+	"github.com/flant/shell-operator/pkg/kube/object_patch"
+	"github.com/flant/shell-operator/pkg/kube_events_manager/types"
+	log "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/utils/pointer"
+
+	v1 "github.com/deckhouse/deckhouse/modules/040-node-manager/hooks/internal/v1"
+)
+
+const (
+	systemReserveMigrationCM = "system-reserve-config-migration"
+	systemReserveMigrationNS = "d8-system"
+)
+
+var _ = sdk.RegisterFunc(&go_hook.HookConfig{
+	OnBeforeHelm: &go_hook.OrderedConfig{Order: 1},
+	Kubernetes: []go_hook.KubernetesConfig{
+		{
+			Name:                         "ngs",
+			ApiVersion:                   "deckhouse.io/v1",
+			Kind:                         "NodeGroup",
+			FilterFunc:                   ngFilter,
+			ExecuteHookOnEvents:          pointer.Bool(false),
+			ExecuteHookOnSynchronization: pointer.Bool(false),
+		},
+		{
+			Name:                         "cm",
+			ApiVersion:                   "v1",
+			Kind:                         "ConfigMap",
+			NameSelector:                 &types.NameSelector{MatchNames: []string{systemReserveMigrationCM}},
+			FilterFunc:                   configMapName,
+			ExecuteHookOnEvents:          pointer.Bool(false),
+			ExecuteHookOnSynchronization: pointer.Bool(false),
+		},
+	},
+}, systemReserve)
+
+type NodeGroup struct {
+	Name string
+}
+
+func ngFilter(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var ng v1.NodeGroup
+	err := sdk.FromUnstructured(obj, &ng)
+	if err != nil {
+		return "", err
+	}
+
+	return &NodeGroup{
+		Name: ng.Name,
+	}, nil
+}
+
+type CM struct {
+	Name string
+}
+
+func configMapName(obj *unstructured.Unstructured) (go_hook.FilterResult, error) {
+	var cm corev1.ConfigMap
+	err := sdk.FromUnstructured(obj, &cm)
+	if err != nil {
+		return "", err
+	}
+
+	return &CM{Name: cm.Name}, nil
+}
+
+func systemReserve(input *go_hook.HookInput) error {
+	if cmSnapshot := input.Snapshots["cm"]; len(cmSnapshot) > 0 {
+		log.Infof("System reserved Nodes are already migrated, skipping...")
+		return nil
+	}
+
+	ngsSnapshot := input.Snapshots["ngs"]
+	for _, ngRaw := range ngsSnapshot {
+		ng := ngRaw.(*NodeGroup)
+		input.PatchCollector.Filter(func(u *unstructured.Unstructured) (*unstructured.Unstructured, error) {
+			objCopy := u.DeepCopy()
+			err := unstructured.SetNestedField(objCopy.Object, "Off", "spec", "kubelet", "resourceReservation", "mode")
```

Comment:
Do we imply that because of 3wm neither argocd nor helm will override the field?

Reply:
Yes, we do.

Reply:
I assume it does not work with werf and argocd out of the box.

Reply:
It's a huge problem then. And I see no solution besides creating a mutating webhook.

But how do ArgoCD and werf play around webhooks? Do they stubbornly re-roll all resources that are changed via a mutating webhoom?

Reply:
https://argo-cd.readthedocs.io/en/release-1.8/user-guide/diffing/

Reply:
It works out of the box. Tested on Argo and werf. Otherwise, it would've been a pain to execute any mutating operations.

Reply:
Live NG:
![image](https://github.com/deckhouse/deckhouse/assets/950313/459dc9fc-338f-4539-93bb-a41acdefa6af)

Desired Argo state:
![image](https://github.com/deckhouse/deckhouse/assets/950313/2ad8ca56-3517-41b2-94a6-7d52a40bbd10)

Sync status is Green.